projects/components/src/table/data/table-cdk-data-source.ts

```diff
@@ -274,19 +283,57 @@ export class TableCdkDataSource implements DataSource<TableRow> {
       return of([]);
     }
 
-    return this.tableDataSourceProvider.data
-      .getData(this.buildRequest(columnConfigs, pageEvent, filters, queryProperties))
-      .pipe(
-        tap(response => this.updatePaginationTotalCount(response.totalCount)),
-        map(response => response.data),
-        map(rows => this.paginateRows(rows, pageEvent)),
-        map(TableCdkRowUtil.buildInitialRowStates),
-        map(rows =>
-          this.rowStateChangeProvider.initialExpandAll && TableCdkRowUtil.isFullyExpandable(rows)
-            ? TableCdkRowUtil.expandAllRows(rows)
-            : rows
-        )
-      );
+    return this.hasCachedRowsForPageEvent(pageEvent)
+      ? this.fetchCachedData(pageEvent)
+      : this.fetchNewData(this.buildRequest(columnConfigs, pageEvent, filters, queryProperties));
+  }
+
+  private hasCachedRowsForPageEvent(pageEvent: PageEvent): boolean {
+    if (this.cachedRows.length < pageEvent.pageSize) {
```

Comment:
should it be `pageEvent.pageSize * pageEvent.pageIndex`?

Reply:
I think the check you are looking for is below.. This line is just a quick sanity check to fail early if we don't even have as many results cached as we have requested.